File: templates/cookbook/multiple-sessions.ts
/**
 * multiple-sessions.ts — CopilotForge Cookbook Recipe
 *
 * WHAT THIS DOES:
 *   Demonstrates how to manage multiple independent Copilot conversations
 *   simultaneously. Each session has its own context, model, and permissions.
 *   Useful for multi-user applications or multi-task workflows.
 *
 * WHEN TO USE THIS:
 *   - Building a multi-user chat app where each user needs their own session
 *   - Running parallel tasks with different AI models (e.g., A/B testing)
 *   - Isolating conversations by context (e.g., Python help vs. TypeScript help)
 *
 * HOW TO RUN:
 *   1. npm install @github/copilot-sdk
 *   2. Set GITHUB_TOKEN: export GITHUB_TOKEN="{{github_token}}"
 *   3. npx ts-node multiple-sessions.ts
 *
 * PREREQUISITES:
 *   - Node.js 18+
 *   - TypeScript 5+
 *   - GitHub Copilot SDK access
 *   - Valid GitHub token with Copilot access
 *
 * EXPECTED OUTPUT:
 *   [Session {{session_name_1}}] Created with model {{model_name}}
 *   [Session {{session_name_2}}] Created with model {{model_name}}
 *   [Session {{session_name_3}}] Created with model {{alternative_model}}
 *   {{language_1}}: {{sample_response_1}}
 *   {{language_2}}: {{sample_response_2}}
 *   {{language_3}}: {{sample_response_3}}
 *   [Cleanup] All 3 sessions destroyed
 */

// --- Imports ---
// All imports are included — no "install this separately" surprises.
import { randomUUID } from "node:crypto";

// --- Types ---

/** Configuration for creating a new session. */
interface SessionConfig {
  id?: string; // Optional custom ID (e.g., "user-123-chat")
  model?: string; // Model to use (e.g., "gpt-5", "claude-sonnet-4.5")
  systemPrompt?: string; // Initial instruction for the session
  metadata?: Record<string, unknown>; // Custom metadata for tracking
}

/** Represents a single Copilot session. */
interface CopilotSession {
  id: string;
  model: string;
  createdAt: Date;
  lastActivity: Date;
  metadata: Record<string, unknown>;
  messageCount: number;
}

// --- Multi-Session Manager ---

/**
 * Manages multiple independent Copilot sessions.
 * Each session has its own conversation history and can use a different model.
 *
 * Usage:
 *   const manager = new MultiSessionManager();
 *   await manager.start();
 *   const sessionId = await manager.createSession({ model: "{{model_name}}" });
 *   const response = await manager.sendMessage(sessionId, "Hello!");
 *   await manager.destroySession(sessionId);
 *   await manager.stop();
 */
class MultiSessionManager {
  private sessions: Map<string, CopilotSession> = new Map();
  private isStarted: boolean = false;

  /**
   * Initializes the Copilot SDK client.
   * Must be called before creating sessions.
   */
  async start(): Promise<void> {
    if (this.isStarted) {
      console.warn("[Manager] Already started");
      return;
    }

    // TODO: Replace with actual CopilotClient initialization:
    //
    //   import { CopilotClient } from "@github/copilot-sdk";
    //   this.client = new CopilotClient();
    //   await this.client.start();

    this.isStarted = true;
    console.log("[Manager] Started");
  }

  /**
   * Creates a new session with optional custom ID and model.
   * If no ID is provided, generates a random UUID.
   *
   * Returns the session ID (useful if auto-generated).
   */
  async createSession(config: SessionConfig = {}): Promise<string> {
    if (!this.isStarted) {
      throw new SessionError("Manager not started. Call start() first.");
    }

    const sessionId = config.id ?? randomUUID();
    const model = config.model ?? "{{default_model}}";

    // Check if session ID already exists.
    if (this.sessions.has(sessionId)) {
      throw new SessionError(
        `Session ${sessionId} already exists. Use a different ID or destroy the existing session.`
      );
    }

    // TODO: Replace with actual Copilot SDK session creation:
    //
    //   const sdkSession = await this.client.createSession({
    //     sessionId: sessionId,
    //     onPermissionRequest: approveAll,
    //     model: model,
    //   });
    //
    //   if (config.systemPrompt) {
    //     await sdkSession.sendAndWait({ prompt: config.systemPrompt });
    //   }

    const session: CopilotSession = {
      id: sessionId,
      model: model,
      createdAt: new Date(),
      lastActivity: new Date(),
      metadata: config.metadata ?? {},
      messageCount: 0,
    };

    this.sessions.set(sessionId, session);
    console.log(`[Session ${sessionId}] Created with model ${model}`);

    return sessionId;
  }

  /**
   * Sends a message to a specific session and waits for the response.
   *
   * Throws if the session doesn't exist.
   */
  async sendMessage(sessionId: string, prompt: string): Promise<string> {
    const session = this.getSessionOrThrow(sessionId);

    // TODO: Replace with actual SDK call:
    //
    //   const sdkSession = this.sdkSessions.get(sessionId);
    //   const response = await sdkSession.sendAndWait({ prompt });
    //   return response.content;

    // Placeholder: echo back for demonstration.
    const response = `[Echo from ${sessionId}] Received: "${prompt}"`;

    session.messageCount++;
    session.lastActivity = new Date();

    return response;
  }

  /**
   * Destroys a specific session and cleans up resources.
   * Safe to call multiple times — silently ignores missing sessions.
   */
  async destroySession(sessionId: string): Promise<void> {
    if (!this.sessions.has(sessionId)) {
      console.warn(`[Session ${sessionId}] Not found (may already be destroyed)`);
      return;
    }

    // TODO: Replace with actual SDK cleanup:
    //
    //   const sdkSession = this.sdkSessions.get(sessionId);
    //   await sdkSession.destroy();
    //   this.sdkSessions.delete(sessionId);

    const session = this.sessions.get(sessionId)!;
    console.log(`[Session ${sessionId}] Destroyed after ${session.messageCount} messages`);
    this.sessions.delete(sessionId);
  }

  /**
   * Lists all active sessions with their metadata.
   * Useful for debugging or building a session management UI.
   */
  listSessions(): CopilotSession[] {
    return Array.from(this.sessions.values());
  }

  /**
   * Destroys all active sessions and shuts down the SDK client.
   * Should be called when your application exits.
   */
  async stop(): Promise<void> {
    if (!this.isStarted) {
      console.warn("[Manager] Not started");
      return;
    }

    const sessionCount = this.sessions.size;

    // Destroy all sessions.
    for (const sessionId of Array.from(this.sessions.keys())) {
      await this.destroySession(sessionId);
    }

    // TODO: Replace with actual client shutdown:
    //
    //   await this.client.stop();

    this.isStarted = false;
    console.log(`[Manager] Stopped after cleaning up ${sessionCount} session(s)`);
  }

  // --- Private helpers ---

  private getSessionOrThrow(sessionId: string): CopilotSession {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new SessionError(
        `Session ${sessionId} not found. It may have been destroyed or never created.`
      );
    }
    return session;
  }
}

// --- Error handling ---

/** Custom error class for session-related failures. */
class SessionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SessionError";
  }
}

// --- Example usage ---

async function main() {
  const manager = new MultiSessionManager();

  try {
    // Initialize the manager.
    await manager.start();

    // Create three sessions for different {{project_context}}.
    // Each has its own context and can use a different model.
    const session1 = await manager.createSession({
      id: "{{session_name_1}}",
      model: "{{model_name}}",
      systemPrompt: "{{system_prompt_1}}",
      metadata: { {{metadata_key_1}}: "{{metadata_value_1}}" },
    });

    const session2 = await manager.createSession({
      id: "{{session_name_2}}",
      model: "{{model_name}}",
      systemPrompt: "{{system_prompt_2}}",
      metadata: { {{metadata_key_2}}: "{{metadata_value_2}}" },
    });

    const session3 = await manager.createSession({
      id: "{{session_name_3}}",
      model: "{{alternative_model}}",
      systemPrompt: "{{system_prompt_3}}",
      metadata: { {{metadata_key_3}}: "{{metadata_value_3}}" },
    });

    // Send messages to each session independently.
    // These run in parallel — no cross-talk between sessions.
    const [response1, response2, response3] = await Promise.all([
      manager.sendMessage(session1, "{{prompt_1}}"),
      manager.sendMessage(session2, "{{prompt_2}}"),
      manager.sendMessage(session3, "{{prompt_3}}"),
    ]);

    console.log("{{context_1}}:", response1);
    console.log("{{context_2}}:", response2);
    console.log("{{context_3}}:", response3);

    // List all active sessions.
    const activeSessions = manager.listSessions();
    console.log(`\n[Manager] ${activeSessions.length} active sessions:`);
    activeSessions.forEach((s) => {
      console.log(
        `  - ${s.id}: ${s.messageCount} messages, last active ${s.lastActivity.toLocaleTimeString()}`
      );
    });

    // Clean up: destroy all sessions.
    await manager.destroySession(session1);
    await manager.destroySession(session2);
    await manager.destroySession(session3);

    console.log("[Cleanup] All sessions destroyed");
  } catch (error) {
    if (error instanceof SessionError) {
      // Handle session-specific errors (not found, already exists).
      console.error("Session error:", error.message);
    } else {
      // Unexpected errors — log and re-throw.
      console.error("Unexpected error:", error);
      throw error;
    }
  } finally {
    // Always shut down the manager when done.
    await manager.stop();
  }
}

// Run the example.
main().catch(console.error);

File: templates/cookbook/multiple-sessions.py
"""
multiple-sessions.py — CopilotForge Cookbook Recipe

WHAT THIS DOES:
    Demonstrates how to manage multiple independent Copilot conversations
    simultaneously. Each session has its own context, model, and permissions.
    Useful for multi-user applications or multi-task workflows.

WHEN TO USE THIS:
    - Building a multi-user chat app where each user needs their own session
    - Running parallel tasks with different AI models (e.g., A/B testing)
    - Isolating conversations by context (e.g., Python help vs. Rust help)

HOW TO RUN:
    1. pip install copilot-sdk   (or your preferred Copilot SDK)
    2. Set GITHUB_TOKEN: export GITHUB_TOKEN="{{github_token}}"
    3. python multiple-sessions.py

PREREQUISITES:
    - Python 3.10+
    - GitHub Copilot SDK access
    - Valid GitHub token with Copilot access

EXPECTED OUTPUT:
    [Manager] Started
    [Session {{session_name_1}}] Created with model {{model_name}}
    [Session {{session_name_2}}] Created with model {{model_name}}
    [Session {{session_name_3}}] Created with model {{alternative_model}}
    {{language_1}}: {{sample_response_1}}
    {{language_2}}: {{sample_response_2}}
    {{language_3}}: {{sample_response_3}}
    [Manager] 3 active sessions
    [Cleanup] All sessions destroyed
"""

from __future__ import annotations

import asyncio
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Optional


# --- Types ---


@dataclass
class CopilotSession:
    """Represents a single Copilot session."""

    id: str
    model: str
    created_at: datetime = field(default_factory=datetime.now)
    last_activity: datetime = field(default_factory=datetime.now)
    metadata: dict[str, Any] = field(default_factory=dict)
    message_count: int = 0


# --- Errors ---


class SessionError(Exception):
    """Raised when a session operation fails (not found, already exists, etc.)."""

    pass


# --- Multi-Session Manager ---


class MultiSessionManager:
    """
    Manages multiple independent Copilot sessions.
    Each session has its own conversation history and can use a different model.

    Usage:
        manager = MultiSessionManager()
        await manager.start()
        session_id = await manager.create_session(model="{{model_name}}")
        response = await manager.send_message(session_id, "Hello!")
        await manager.destroy_session(session_id)
        await manager.stop()
    """

    def __init__(self) -> None:
        self._sessions: dict[str, CopilotSession] = {}
        self._is_started: bool = False

    async def start(self) -> None:
        """
        Initializes the Copilot SDK client.
        Must be called before creating sessions.
        """
        if self._is_started:
            print("[Manager] Already started")
            return

        # TODO: Replace with actual CopilotClient initialization:
        #
        #   from copilot_sdk import CopilotClient
        #   self._client = CopilotClient()
        #   await self._client.start()

        self._is_started = True
        print("[Manager] Started")

    async def create_session(
        self,
        id: Optional[str] = None,
        model: str = "{{default_model}}",
        system_prompt: Optional[str] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> str:
        """
        Creates a new session with optional custom ID and model.
        If no ID is provided, generates a random UUID.

        Returns the session ID (useful if auto-generated).
        """
        if not self._is_started:
            raise SessionError("Manager not started. Call start() first.")

        session_id = id or str(uuid.uuid4())

        # Check if session ID already exists.
        if session_id in self._sessions:
            raise SessionError(
                f"Session {session_id} already exists. Use a different ID or destroy the existing session."
            )

        # TODO: Replace with actual Copilot SDK session creation:
        #
        #   sdk_session = await self._client.create_session(
        #       session_id=session_id,
        #       on_permission_request=approve_all,
        #       model=model,
        #   )
        #
        #   if system_prompt:
        #       await sdk_session.send_and_wait(prompt=system_prompt)

        session = CopilotSession(
            id=session_id,
            model=model,
            metadata=metadata or {},
        )

        self._sessions[session_id] = session
        print(f"[Session {session_id}] Created with model {model}")

        return session_id

    async def send_message(self, session_id: str, prompt: str) -> str:
        """
        Sends a message to a specific session and waits for the response.

        Raises SessionError if the session doesn't exist.
        """
        session = self._get_session_or_raise(session_id)

        # TODO: Replace with actual SDK call:
        #
        #   sdk_session = self._sdk_sessions[session_id]
        #   response = await sdk_session.send_and_wait(prompt=prompt)
        #   return response.content

        # Placeholder: echo back for demonstration.
        response = f'[Echo from {session_id}] Received: "{prompt}"'

        session.message_count += 1
        session.last_activity = datetime.now()

        return response

    async def destroy_session(self, session_id: str) -> None:
        """
        Destroys a specific session and cleans up resources.
        Safe to call multiple times — silently ignores missing sessions.
        """
        if session_id not in self._sessions:
            print(f"[Session {session_id}] Not found (may already be destroyed)")
            return

        # TODO: Replace with actual SDK cleanup:
        #
        #   sdk_session = self._sdk_sessions[session_id]
        #   await sdk_session.destroy()
        #   del self._sdk_sessions[session_id]

        session = self._sessions[session_id]
        print(f"[Session {session_id}] Destroyed after {session.message_count} messages")
        del self._sessions[session_id]

    def list_sessions(self) -> list[CopilotSession]:
        """
        Lists all active sessions with their metadata.
        Useful for debugging or building a session management UI.
        """
        return list(self._sessions.values())

    async def stop(self) -> None:
        """
        Destroys all active sessions and shuts down the SDK client.
        Should be called when your application exits.
        """
        if not self._is_started:
            print("[Manager] Not started")
            return

        session_count = len(self._sessions)

        # Destroy all sessions.
        for session_id in list(self._sessions.keys()):
            await self.destroy_session(session_id)

        # TODO: Replace with actual client shutdown:
        #
        #   await self._client.stop()

        self._is_started = False
        print(f"[Manager] Stopped after cleaning up {session_count} session(s)")

    # --- Private helpers ---

    def _get_session_or_raise(self, session_id: str) -> CopilotSession:
        session = self._sessions.get(session_id)
        if session is None:
            raise SessionError(
                f"Session {session_id} not found. It may have been destroyed or never created."
            )
        return session


# --- Example usage ---


async def main() -> None:
    manager = MultiSessionManager()

    try:
        # Initialize the manager.
        await manager.start()

        # Create three sessions for different {{project_context}}.
        # Each has its own context and can use a different model.
        session1 = await manager.create_session(
            id="{{session_name_1}}",
            model="{{model_name}}",
            system_prompt="{{system_prompt_1}}",
            metadata={"{{metadata_key_1}}": "{{metadata_value_1}}"},
        )

        session2 = await manager.create_session(
            id="{{session_name_2}}",
            model="{{model_name}}",
            system_prompt="{{system_prompt_2}}",
            metadata={"{{metadata_key_2}}": "{{metadata_value_2}}"},
        )

        session3 = await manager.create_session(
            id="{{session_name_3}}",
            model="{{alternative_model}}",
            system_prompt="{{system_prompt_3}}",
            metadata={"{{metadata_key_3}}": "{{metadata_value_3}}"},
        )

        # Send messages to each session independently.
        # These run in parallel — no cross-talk between sessions.
        response1, response2, response3 = await asyncio.gather(
            manager.send_message(session1, "{{prompt_1}}"),
            manager.send_message(session2, "{{prompt_2}}"),
            manager.send_message(session3, "{{prompt_3}}"),
        )

        print("{{context_1}}:", response1)
        print("{{context_2}}:", response2)
        print("{{context_3}}:", response3)

        # List all active sessions.
        active_sessions = manager.list_sessions()
        print(f"\n[Manager] {len(active_sessions)} active sessions:")
        for session in active_sessions:
            print(
                f"  - {session.id}: {session.message_count} messages, "
                f"last active {session.last_activity.strftime('%H:%M:%S')}"
            )

        # Clean up: destroy all sessions.
        await manager.destroy_session(session1)
        await manager.destroy_session(session2)
        await manager.destroy_session(session3)

        print("[Cleanup] All sessions destroyed")

    except SessionError as e:
        # Handle session-specific errors (not found, already exists).
        print(f"Session error: {e}")

    except Exception as e:
        # Unexpected errors — log and re-raise.
        print(f"Unexpected error: {e}")
        raise

    finally:
        # Always shut down the manager when done.
        await manager.stop()


if __name__ == "__main__":
    asyncio.run(main())
